returnn/tf/layers/basic.py

```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
+    :param int|None seed: random seed
+    :param kwargs:
+    """
+    from returnn.tf.util.data import DimensionTag
+    super(RandIntLayer, self).__init__(**kwargs)
+
+    shape_parsed = []
+    batch_dim_axes = []
+    dyn_axes_sizes = []
+    for ax, s in enumerate(shape):
+      if isinstance(s, int):
+        shape_parsed.append(s)
+      elif isinstance(s, DimensionTag):
+        if s.kind == DimensionTag.Types.Batch:
+          batch_dim_axes.append(ax)
+          shape_parsed.append(self.get_batch_dim())
+        elif isinstance(s.dimension, int):
+          shape_parsed.append(s.dimension)
+        elif s.dimension is None:
+          assert s.dyn_size is not None
+          shape_parsed.append(tf.reduce_max(s.dyn_size))
+          dyn_axes_sizes.append(tf.fill(tf.shape(s.dyn_size), shape_parsed[-1]))
+        else:
+          assert False
+      else:
+        raise NotImplementedError
+    shape_parsed = tuple(shape_parsed)
+
+    if batch_dim_axes:
+      assert len(batch_dim_axes) == 1
+      self.output.size_placeholder = {i: size_placeholder for i, size_placeholder in enumerate(dyn_axes_sizes)}
+    else:
+      self.output.size_placeholder = {}
+
+    self.output.placeholder = tf.random.uniform(shape_parsed, minval=minval, maxval=maxval, dtype=dtype, seed=seed)
+
+  @classmethod
+  def transform_config_dict(cls, d, network, get_layer):
+    """
+    :param dict[str] d:
+    :param returnn.tf.network.TFNetwork network:
+    :param (str)->LayerBase get_layer:
+    """
+    d.setdefault("from", [])
+    super(RandIntLayer, cls).transform_config_dict(d, network=network, get_layer=get_layer)
+
+  @classmethod
+  def get_out_data_from_opts(cls, name, shape, maxval, minval=0, dtype="int32", **kwargs):
+    """
+    :param str name:
+    :param tuple[DimensionTag|int]|list[DimensionTag|int] shape: desired shape of output tensor
+    :param int maxval: upper bound (exclusive) on range of random values
+    :param int minval: lower bound (inclusive) on range of random values
+    :param str dtype: type of the output. For random ints, int32 and int64 make sense, but could also be floats
+    :param kwargs:
+    :rtype: Data
+    """
+    from returnn.tf.util.data import DimensionTag, NotSpecified
+
+    shape_parsed = []
+    batch_dim_axis = None
+    time_dim_axis = None
+    feature_dim_axis = None
+    for ax, s in enumerate(shape):
+      if isinstance(s, int):
+        shape_parsed.append(s)
+        feature_dim_axis = ax
+      elif isinstance(s, DimensionTag):
+        if s.kind == DimensionTag.Types.Batch:
+          batch_dim_axis = ax
+        elif isinstance(s.dimension, int):
+          shape_parsed.append(s.dimension)
+          feature_dim_axis = ax
+        elif s.dimension is None:
+          shape_parsed.append(None)
+          time_dim_axis = ax
+        else:
+          assert False
+      else:
+        raise NotImplementedError
+    shape_parsed = tuple(shape_parsed)
```

Comment:
You have duplicate code here. Put that into a (class) helper function.

Reply:
Ah sorry, I see now it's not exactly the same.
Then leave it as it is. A helper function would just complicate it more than it helps.